core/shared/src/main/scala/fs2/async/async.scala

```diff
@@ -104,21 +105,37 @@ package object async {
   def fork[F[_], A](f: F[A])(implicit F: Effect[F], ec: ExecutionContext): F[Unit] =
     F.liftIO(F.runAsync(F.shift *> f) { _ => IO.unit })
 
+  /**
+   * Like `unsafeRunSync` but execution is shifted to the supplied execution context.
+   * This method returns immediately after submitting execution to the execution context.
+   */
+  def unsafeRunAsync[F[_], A](fa: F[A])(f: Either[Throwable, A] => IO[Unit])(implicit F: Effect[F], ec: ExecutionContext): Unit =
+    F.runAsync(F.shift(ec) *> fa)(f).unsafeRunSync
+
   /**
     * Returns an effect that, when run, races evaluation of `fa` and `fb`,
     * and returns the result of whichever completes first. The losing effect
     * continues to execute in the background though its result will be sent
     * nowhere.
    */
-  def race[F[_]: Effect, A, B](fa: F[A], fb: F[B])(implicit ec: ExecutionContext): F[Either[A, B]] =
-    promise[F, Either[A,B]].flatMap { p =>
-      p.race(fa.map(Left.apply), fb.map(Right.apply)) *> p.get
+private[fs2] def race[F[_], A, B](fa: F[A], fb: F[B])(implicit F: Effect[F], ec: ExecutionContext): F[Either[A, B]] =
```

Comment:
Late feedback, does the `race` method need to be private?  It's used from http4s.

https://github.com/http4s/http4s/blob/0d1ebe8cfbe439703dd189dd0af663d8d9bbed16/server/src/main/scala/org/http4s/server/middleware/Timeout.scala#L26

Reply:
On second thoughts, I think it's correct for this to be private given the `unsafe` calls it contains

Reply:
No that isn't the reason. Those `unsafeRunAsync` are wrapped in `F`, so the method is still perfectly pure. The reason is private is because it can't guarantee resource safety for the losing `F`. We will consider making it public again I think if  `http4s` needs it, I'll think about it

Reply:
https://github.com/http4s/http4s/issues/1583